tools/state-viewer/src/state_parts.rs

```diff
@@ -341,11 +343,51 @@ fn dump_state_parts(
             .obtain_state_part(shard_id, &sync_hash, &state_root, PartId::new(part_id, num_parts))
             .unwrap();
         part_storage.write(&state_part, part_id, num_parts);
-        tracing::info!(target: "state-parts", part_id, part_length = state_part.len(), elapsed_sec = timer.elapsed().as_secs_f64(), "Wrote a state part");
+        let elapsed_sec = timer.elapsed().as_secs_f64();
+        let (first_sr, last_sr) = analyze_state_part(&state_root, &state_part);
+        let (first_sr, first_sr_key_len) =
+            first_sr.map_or((None, None), |sr| (Some(sr.0), Some(sr.1)));
+        let (last_sr, last_sr_key_len) =
+            last_sr.map_or((None, None), |sr| (Some(sr.0), Some(sr.1)));
+        tracing::info!(
+            target: "state-parts",
+            part_id,
+            part_length = state_part.len(),
+            elapsed_sec,
+            ?first_sr,
+            ?first_sr_key_len,
+            ?last_sr,
+            ?last_sr_key_len,
+            "Wrote a state part");
     }
     tracing::info!(target: "state-parts", total_elapsed_sec = timer.elapsed().as_secs_f64(), "Wrote all requested state parts");
 }
 
+/// Returns the first and the last `StateRecord`s encountered while iterating over a state part.
+fn analyze_state_part(
+    state_root: &StateRoot,
+    data: &[u8],
+) -> (Option<(StateRecord, usize)>, Option<(StateRecord, usize)>) {
+    let trie_nodes = BorshDeserialize::try_from_slice(data).unwrap();
+    let trie = Trie::from_recorded_storage(PartialStorage { nodes: trie_nodes }, *state_root);
+
+    let mut first_sr = None;
+    let mut last_sr = None;
+    for item in trie.iter().unwrap() {
+        if let Ok((key, value)) = item {
+            let key_len = key.len();
+            if let Some(sr) = StateRecord::from_raw_key_value(key, value) {
```

Comment:
Totally fine if this is what's actually useful, but is printing out the first and last state record exactly required? looks like it's just to print it out as debugging info, so it feels to me like quite a bit of work to call `StateRecord::from_raw_key_value ` on every node just to print only that out in the end. For the debugging purposes here would it do to just call `Trie::print_recursive()` w/ a low max depth?

Reply:
After trying this code out, I simplified:
* Removed the node key length
* Print only the first state record, not the last

`print_recursive()` would be good, but I don't want too much output to stderr.